src/ipc/ipc-helper.c

```diff
@@ -180,12 +179,12 @@ int comp_verify_params(struct comp_dev *dev, uint32_t flag,
 int comp_buffer_connect(struct comp_dev *comp, uint32_t comp_core,
 			struct comp_buffer *buffer, uint32_t dir)
 {
+	struct comp_dev *cd = buffer_get_comp(buffer, dir);
 	int ret;
 
 	/* check if it's a connection between cores */
-	if (buffer->core != comp_core) {
+	if (cd && cd->ipc_config.core != comp_core) {
```

Comment:
you could add a comment, that since this is called twice for each buffer - once for up- and once for downstream, during the second connection we can check the already connected direction to compare cores.